coalib/bearlib/aspects/__init__.py

```diff
@@ -0,0 +1,284 @@
+    ... ]
+
+    >>> Aspects.Redundancy.Clone.settings[0].default
+    30
+    >>> Aspects.Redundancy.Clone.Logic.settings[0].default
+    30
+
+    If you don't want this, you can override the settings the particular
+    sub-aspect:
+
+    >>> Aspects.Redundancy.Clone.Logic.settings = [
+    ...     AspectSetting("min_clone_token",
+    ...                   "The number of tokens that have to be "
+    ...                   "equal for it to be detected as a code clone",
+    ...                   int, default=40)
+    ... ]
+
+    >>> Aspects.Redundancy.Clone.settings[0].default
+    30
+    >>> Aspects.Redundancy.Clone.Logic.settings[0].default
+    40
+
+    The user can use the qualifying name according to PEP 3155 for selecting
+    and unselecting aspects (see <https://www.python.org/dev/peps/pep-3155/>):
+
+    >>> Aspects.Redundancy.Clone.__qualname__
+    'Redundancy.Clone'
+    '''
+
+    def __init__(self, name="Aspects", parent=None):
+        """
+        Instantiates the ``Aspect`` object.
+
+        :param name:
+            Name of the aspect.
+        :param parent:
+            A reference to the parent object for this aspect. For first-level
+            aspects, this is the root aspect.
+        """
+        self.__name__ = name
+        self.docs = AspectDocumentation("", "", "", "")
+        self.subaspects = {}
+        self.parent = parent
+        self.__settings__ = []
+
+        current = self
+        quals = []
+
+        while current is not None and current.__name__ != "Aspects":
+            quals.append(current.__name__)
+            current = current.parent
+        quals.reverse()
+
+        self.__qualname__ = ".".join(quals)
+
+    def new_subaspect(self, name):
+        """
+        Creates a sub-aspect.
+
+        :param name:
+            Name of the sub-aspect.
+        """
+        self.subaspects[name] = Aspect(name, self)
+
+    def __getattr__(self, subaspect):
+        """
+        Returns the sub-aspect.
+
+        :param subaspect:
+            Name of the sub-aspect. This should have been instantiated
+            previously using the ``new_subaspect`` method.
+        :return:
+            An ``Aspect`` object corresponding to the sub-aspect. If the
+            sub-aspect is not found, ``None`` is returned.
+        """
+        if subaspect in self.__dict__["subaspects"]:
+            return self.__dict__["subaspects"][subaspect]
+        return None
+
+    @property
+    def settings(self):
+        current = self
+        taken_settings = set()
+        result = []
+
+        # Recursively go up and collect settings (the most local setting
+        # overrides previous values)
+        while current is not None:
+            for setting in current.__settings__:
+                if setting.name not in taken_settings:
+                    taken_settings.add(setting.name)
+                    result.append(setting)
+            current = current.parent
```

Comment:
Because we want the closest settings to override higher level settings. IIUC with what you suggest, we would have multiple copies for the same settings (with possibly different values).